hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/MappingModelCreationHelper.java

```diff
@@ -814,6 +859,61 @@ public static void interpretKeyDescriptor(
 		}
 	}
 
+	public static EmbeddedForeignKeyDescriptor buildEmbeddedForeignKeyDescriptor(
+			EmbeddableValuedModelPart fkTarget,
+			AbstractAttributeMapping attributeMapping,
+			Value bootValueMapping,
+			Dialect dialect,
+			MappingModelCreationProcess creationProcess) {
+		final List<String> keyColumnExpressions = new ArrayList<>();
+		bootValueMapping.getColumnIterator().forEachRemaining(
+				column ->
+						keyColumnExpressions.add( column.getText( dialect ) ) );
+
+		final List<String> targetColumnExpressions = new ArrayList<>();
+		fkTarget.getMappedColumnExpressions().forEach(
+				column ->
+						targetColumnExpressions.add( column ) );
+
```

Comment:
seems we can specify ArrayList size in constructor above to avoid performance and memory overhead, though not a big deal.